### ISSUE TITLE
Redirect URI error when running demo code in Azure Web App Service. Like #465

### ISSUE THREAD

**Issue by twofingerrightclick** [Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2#499]
I have the similar problem to #465 (couldn't reopen issue)  (except I am using a regular windows 64 Azure web app service). I can authenticate when running locally, but as soon as I try using the app registrations authentication provider with an Azure Web App service I get the redirect issue: 
![image](https://user-images.githubusercontent.com/11265760/121756477-053f2f00-cacf-11eb-9b60-5e2a2cdbe5fe.png)

My app registration redirects:
![image](https://user-images.githubusercontent.com/11265760/121756555-5c450400-cacf-11eb-8695-b397ac8bb8d3.png)

When I look at the header of the login when I get the redirect error I see that its trying to redirect to `https://***/.auth/login/aad/callback` (the actual request url: https://login.microsoftonline.com/***tenantID***/oauth2/authorize?response_type=code+id_token&redirect_uri=https%3A%2F%2F**appServiceName**%2F.auth%2Flogin%2Faad%2Fcallback&client_id=******&scope=openid+profile+email) , and if I add that uri to the redirect list in the app registration's Authentication, I no longer get the redirect error, but instead get back to my azure app and it throws a server error (probably because there is no framework code for handling that).  --- the request url when using the localhost app works because the `redirect_uri=` "/signin-oidc " 
![image](https://user-images.githubusercontent.com/11265760/121758404-12135100-cad6-11eb-91cf-d36e9467e900.png)

I also observe that in creating an app registration with identity from the app service (so not creating the registration before hand and then adding it to the web app service, but using the tool in the web app service) it will automatically only add the one redirect uri to the registration's redirect uri's : `https://***/.auth/login/aad/callback` . So why is my azure web app service wanting that to be the redirect, but the examples in this repo all say to use `/signin-oidc` ? I am trying to figured out what paradigms currently exist for AAD auth, as there appears to be many in use and sounds like there is a new "EasyAuth" now coming into play as well... (e.g this [Microsoft quick start doc](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app#add-a-redirect-uri) mentions other uri's that based of the context given in the doc are for web apps... I can't find an authoritative list of uri's and what cases they are to be used. )

### TLDR 
I am wondering if the default `Issuer URL` in "name of app service" -> Authentication-> (click edit on the current provider)->Edit identity provider  is the problem as this is the only difference between the localhost instance of the demo site and the Azure Web app service instance.... Here is what I get:

![image](https://user-images.githubusercontent.com/11265760/121757137-636d1180-cad1-11eb-8c8e-fa33ab092f4a.png)

_Originally posted by @twofingerrightclick in https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/issues/465#issuecomment-859954898_

I understand this isn't an issue with this repo, but it is one of the few places I've found good documentation for AAD auth with actual example code.  How can this be configured to work with an Azure App Service Web app?

**Comment by jmprieur** [comment]
@twofingerrightclick.
If you can run it locally and it does all the authentication, you don't need to use App service **authentication**, when you deploy to App services. (you can, but don't need to). Depending on whether you enable or not App service authentication, the redirect URI will be different:

 Scenario |  Redirect URI
-----------   | -----------  
Run on your developer box with IIS |  ` https://localhost:44321/signin-oidc`
Run on your developer box with Kestrel profile |  ` https://localhost:5001/signin-oidc`
Deployed to app service without app service authentication | `https://appServiceBaseUri/signin-oidc`
Deployed to app service **with** app service authentication | `https://appServiceBaseUri/.auth/login/aad/callback`

Depending on the scenarios you want to run, you should add the corresponding redirect URI to the app registration

**Comment by jmprieur** [comment]
I've also updated the doc.

**Comment by twofingerrightclick** [comment]
@jmprieur Thank you for clarifying! I now understand more that App Service Authentication is to be used without  the "code behind" that is in this demonstrated in this repo, but it requires third party code if you want to add role-based authorization at this point (https://github.com/MaximRouiller/MaximeRouiller.Azure.AppService.EasyAuth).
I was mislead to enabling App Service Authentication because when I just deploy the app to the App Service I was getting a: 
![image](https://user-images.githubusercontent.com/11265760/121784238-b09bc280-cb67-11eb-8401-f8c129911174.png) 
and so thought I had to go the extra step of enabling it and linking it to the App registration. So with just a App registration correctly configured with both localhost and azurewebsites redirects, and with the localhost instance working, when I publish it to an app service I get the "...0000000 is not a valid application identifier". 
![image](https://user-images.githubusercontent.com/11265760/121784379-98787300-cb68-11eb-989b-5ee715af4842.png)

**Comment by jmprieur** [comment]
@twofingerrightclick, normally MIcrosoft.Identity.Web detects that you detected to app service and enabled authentication (but that doesn't prevent you from adding the right redirect URI).
It works well for simple scenarios, but less for complex Web apps that call multiple web APIs and require incremental consent.

**Comment by jmprieur** [comment]
BTW, if you haven't, you might want to update the version of Microsoft.Identity.Web to 1.13.0 (that was released yesterday. We also updated this sample to that version yesterday fixing a couple of issues with Easy auth, who took a breaking change between v1.0 and v2.0)

**Comment by twofingerrightclick** [comment]
@jmprieur Got 1.13.0. 
To clarify: I want to use Microsoft.Identity.Web like so: 
![image](https://user-images.githubusercontent.com/11265760/121786371-53f2d480-cb74-11eb-88c1-35b4532b24fe.png)
and so in the web app service I should have?: 
![image](https://user-images.githubusercontent.com/11265760/121786407-7c7ace80-cb74-11eb-9667-3fbdb8a1f022.png)
(I also just tried publishing to a brand new web app service in which I hadn't previously added Authentication to, and I still get the ".00000 is not" error)
If so, will try to figure out why I get the "...0000000 is not a valid application identifier". I for sure have the "AzureAd" in the configuration file (I see it in the Configuration provider coming from the appsettings.json when debugging). Is Microsoft.Identity.Web realizing its in a Web App Service Environment and not using it like it does when I run it locally?

**Comment by jmprieur** [comment]
@twofingerrightclick 
There is a tutorial that will tell you all you want to know:

https://docs.microsoft.com/en-us/azure/app-service/scenario-secure-app-authentication-app-service

**Comment by twofingerrightclick** [comment]
Thanks. But I've read that one several times :) ... But I will close the issue and comment later here later with what I figure out. I am probably missing something simple. A coworker is writing a blog about how to use the Web App Service authentication with Authorization, and I might write one about how to use the old approach of Authentication that's been around since Aps.net 2 when with Web App Services today (that's what I am trying to do).
